python/social_bot/envs/grocery_ground.py

```diff
@@ -608,9 +640,9 @@ def step(self, action):
             action_ctrl = action
         controls = np.clip(action_ctrl, -1.0, 1.0) * self._agent_control_range
         controls = dict(zip(self._agent_joints, controls))
-        teacher_action = self._teacher.teach(sentence)
         self._agent.take_action(controls)
         self._world.step(self._sub_steps)
+        teacher_action = self._teacher.teach(sentence)
```

Comment:
Similar to above comment. I'm not very sure here.

Reply:
Right, it should be after step()

Reply:
The problem of placing it after world.step() is that if teacher changes the state of the world, the agent may not see it immediately. And of course, placing it before world.step() has its own problem.

Reply:
Maybe we can separate the original teacher.teach() to 2 functions like teach() and give_reward(), and place teach() before world.step()?

Reply:
Oh, seems the separation won't help. Agent get observation from last step 
```
step(agent_action, agent_sentence):
    agent.take_action(agent_action)
    teacher_sentence = teacher.teach(agent_sentence) # give sentence and might do some changes to world
    world.step()
    reward = teacher.give_reward() # give reward
    agent_observation = teacher_sentence + observation_from_world
    return agent_observation, reward
```

```
step(agent_action, agent_sentence):
    agent.take_action(agent_action)
    world.step()
    reward, teacher_sentence = teacher.teach(agent_sentence) # give reward, sentence and might do some changes to world
    agent_observation = teacher_sentence + observation_from_world
    return agent_observation, reward
```

Reply:
Just thinking out loud, shouldn't step() first let agent observe and take action, and world.step(), and then teacher teach?  Reward is part of world.step()?

If agent action happens before observation, then the first action will be kind of random?

Probably doesn't affect performance that much though.

Reply:
Teacher will do first teach() in reset(), then agent will get the observation from reset and generate actions to call env.step()

Reply:
So we can keep it this way, put teach() before step().

Reply:
One possible way is to put teacher.teach() between two world.step() so that teacher can get teach based on the immediate result from the actions of the learner and give reward and change the world state if necessary.
```python
    self._agent.take_action(controls)
    self._world.step(self._sub_steps/2, skip_rendering)
    teacher_action = self._teacher.teach(sentence)
    self._world.step(self._sub_steps/2)
```
But I am not absolutely sure about whether this change won't complicate the logic in teacher or not.
```

Reply:
I think JT's second option is more natural?
If teacher wants to change world state (e.g. switch to a different task), the teacher can do a world.reset() inside teach()?  If the teacher is continuing the same task, then maybe the teacher need to see the effect of agent's action before giving the instruction, and the teacher may also change the env, so it's natural that the agent will delay seeing the effect of the teacher changing the world by 1 world.step().

Does that make sense?


```
step(agent_action, agent_sentence):
    agent.take_action(agent_action)
    world.step()
    reward, teacher_sentence = teacher.teach(agent_sentence) # give reward, sentence and might do some changes to world
    agent_observation = teacher_sentence + observation_from_world
    return agent_observation, reward
```

Reply:
I think we can keep it this simple way for now, If necessary we can change to the half step formula (or add option for it) anytime in the future.